### ISSUE TITLE
Design: Selective VLANs on the Root Device and Ease of Network Orchestration

### ISSUE THREAD

**Issue by pperiyasamy** [kubevirt/ovs-cni#96]
### **Problem:**

The OVS CNI in its present form supports two types of OVS network attachment definitions (NAD).

1. Single VLAN untagged interface

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata: 
  name: ovs_br-bond_1024
spec:
  config: '{
      "cniVersion": "0.3.1",
      "type": "ovs",
      "bridge": "br-bond",
      "vlan": 1024
}'
```

2. Transparent trunk 

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata: 
  name: ovs_br-int_transparent_trunk
spec:
  config: '{
      "cniVersion": "0.3.1",
      "type": "ovs",
      "bridge": "br-bond",
}'
```
Note: The absence of the “vlan” field implies transparent access to all VLANs present on the specified bridge and underlying root device.

In case of Transparent trunk scenario, the root device is entirely given to application(s) running in the pod container(s) and can not used by infrastructure or primary networks. Hence there should be a mechanism to use a set of VLANs by secondary networks and remaining VLANs can be used by infrastructure.

There should also ease of network orchestration which allows pod can request for specific VLAN(s) in the allowed VLAN range.  This would reduce number of NAD objects tremendously.

### **Proposal:**

For selective trunk network attachments, enhance the OVS CNI with the following option:

3. Selective trunk 

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata: 
  name: ovs_br-bond_trunk_1-100_200_300
spec:
  config: '{
      "cniVersion": "0.3.1",
      "type": "ovs",
      "bridge": "br-bond",
      “trunk”: "1-100,200,300"
}'
```

The new optional “trunk” attribute enumerates the list of allowed VLAN IDs on the trunk interface. Ranges of VLAN IDs can be specified.

During the ADD command the OVS CNI creates a veth pair, moves one endpoint into the pod and attaches the host endpoint to the specified OVS bridge as access port for the specified VLAN (1), as transparent trunk port (2) or as trunk port with list of allowed VLAN IDs (3).

The above three OVS NAD formats have a well-defined VLAN visibility. If there are many VLANs, the Kubernetes operator would have to create a correspondingly large number of NADs. Especially potential space of the selective trunk NADs is huge.

It would significantly simplify network orchestration if the operator provides generic NADs for a bridge that only specify the mode of the OVS port (untagged access or trunk) and a list of allowed VLAN IDs and leave it to the pod specification to select which of the allowed VLAN IDs to use.
To this end we suggest two more optional fields in the OVS NAD: 

```
mode: 	         access | trunk
allowed-vlans:	 List of allowed VLAN IDs
```

The allowed_vlans parameter ensures that pods cannot access VLANs on the bridge that are used by shall be isolated (e.g. infrastructure or primary networks). 

With this there will be two additional OVS NAD variants:

4. Pod-specified untagged single VLAN interface

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata: 
  name: ovs_br-bond_access
spec:
  config: '{
      "cniVersion": "0.3.1",
      "type": "ovs",
      "bridge": "br-bond",
      "mode": "access",
      "allowed-vlans": "1-100,200,300"
}'
```

5. Pod-specified selective trunk interface

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata: 
  name: ovs_br-bond_trunk
spec:
  config: '{
      "cniVersion": "0.3.1",
      "type": "ovs",
      "bridge": "br-bond",
      "mode": "trunk",
      "allowed-vlans": "1-100,200,300"
}'
```

The pod specification will then encode the desired VLAN ID(s) as cni-args in its network annotation, e.g.:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    k8s.v1.cni.cncf.io/networks: |
      [
        {
          "name":" ovs_br-bond_access",
           "cni-args": {
              "wanted-vlans": "300"
          }
        }
      ]
```

It will configure the OVS ports for the wanted VLAN ID(s), rejecting or dropping VLAN IDs that are not in the allowed_vlans list of the NAD.

**Comment by EdDev** [comment]
The 1st part, to specify vlan ID ranges is a welcome addition.
At the host networking level, the following schema snippet has been choosen: https://github.com/nmstate/nmstate/blob/master/libnmstate/schemas/operational-state.yaml#L335 , you could consider the same for the NAD.

The 2nd part, AFAIU suggests to use the NAD as a semi template and use the pod annotation to complete it. Conceptually, this seems problematic.
A NAD defines a connection of a POD and it fits the notion of a "network". Given a RED network, define a NAD and tell the relevant POD/s to use it. With the suggestion here, this will no longer be defined as such.

Could you please elaborate in what scenario you see the need to create so many access VLAN access NAD/s?

**Comment by JanScheurich** [comment]
The typical use case for hundreds of VLANs are user plane gateway nodes like the 5G Core UPF. These nodes may have to interface hundreds of e.g. enterprise VPNs in the SGi side. Typically a single large pod would use VLAN trunking to handle all relevant VLANs on a single interface and separate the VLANs on application level.

But the application might also chose to split the VLANs over many smaller pods so that each pod handles only a single VLAN. In that case the admin would have to create many NAD CRDs. The NAD "template" as you call it would achieve the goal to let the application create as many network attachments as they want within the setof allowed VLANs that the operator has specified in the NAD.

But I appreciate your concern with regarding the semantic of a NAD. If it should strictly refer to a single specific network we can't leave the choice of the VLAN to the pod spec.

**Comment by dankenigsberg** [comment]
> 
> But the application might also chose to split the VLANs over many smaller pods so that each pod handles only a single VLAN. In that case the admin would have to create many NAD CRDs. The NAD "template" as you call it would achieve the goal to let the application create as many network attachments as they want within the set of allowed VLANs that the operator has specified in the NAD.

In this case, it would feel more natural to me to have an operator that creates NADs based on a template and parameters.

> 
> But I appreciate your concern with regarding the semantic of a NAD. If it should strictly refer to a single specific network we can't leave the choice of the VLAN to the pod spec.

There are legitimate cases where a NAD bundles multiple networks. I would love to see ovs-cni support a `trunk` connection, which lets it connect to multiple VLANs. Per-pod annotation limiting the set of allowed vlans seems a bit awkward (as in, I'm not familiar that it is done like that on physical switch ports), but it can be discussed as a follow-up of normal trunk support.

**Comment by pperiyasamy** [comment]
> I would love to see ovs-cni support a `trunk` connection, which lets it connect to multiple VLANs.

Yes @dankenigsberg we've patch locally and will raise it for review in a couple of days.

**Comment by phoracek** [comment]
Implemented via #97
